File: packages/mirascript/src/helpers/convert/to-number.ts
import type { VmAny } from '../../vm/index.js';
import { VmError } from '../error.js';
import { display } from '../serialize.js';
import { isNaN, NegativeInfinity, NotNumber, PositiveInfinity } from '../utils.js';

/** 将字符串转换为数字 */
function parseNumericLiteral(value: string): number | null {
    if (value === '') return null;
    const ch0 = value.charAt(0);
    if (ch0 < '0' || ch0 > '9') {
        return null;
    }
    // 二进制
    if (/^0[bB][01]+$/.test(value)) {
        const num = Number.parseInt(value.slice(2), 2);
        return isNaN(num) ? null : num;
    }
    // 八进制
    if (/^0[oO][0-7]+$/.test(value)) {
        const num = Number.parseInt(value.slice(2), 8);
        return isNaN(num) ? null : num;
    }
    // 十六进制
    if (/^0[xX][0-9a-fA-F]+$/.test(value)) {
        const num = Number.parseInt(value.slice(2), 16);
        return isNaN(num) ? null : num;
    }
    // 十进制/科学计数法
    if (/^[0-9]+(\.[0-9]+)?([eE][+-]?[0-9]+)?$/.test(value)) {
        const num = Number.parseFloat(value);
        return isNaN(num) ? null : num;
    }
    return null;
}

/** 将字符串转换为数字 */
function stringToNumber(value: string): number | null {
    value = value.trim();
    if (value === '') return null;
    if (value === 'inf' || value === '+inf' || value === 'Infinity' || value === '+Infinity') {
        return PositiveInfinity;
    }
    if (value === '-inf' || value === '-Infinity') {
        return NegativeInfinity;
    }
    if (value === 'nan' || value === 'NaN') {
        return NotNumber;
    }
    if (value.startsWith('-')) {
        const num = parseNumericLiteral(value.slice(1));
        if (num !== null) return -num;
        return null;
    } else if (value.startsWith('+')) {
        return parseNumericLiteral(value.slice(1));
    } else {
        return parseNumericLiteral(value);
    }
}

/** 转换为 number */
export function toNumber<F = undefined>(value: VmAny, fallback?: F): number | Exclude<F, undefined> {
    if (typeof value === 'number') return value;
    if (typeof value == 'boolean') return value ? 1 : 0;
    if (typeof value == 'string') {
        const num = stringToNumber(value);
        if (num !== null) return num;
    }
    if (fallback === undefined) {
        throw new VmError(`Failed to convert value to number: ${display(value)}`, NotNumber);
    }
    return fallback as Exclude<F, undefined>;
}

File: crates/python/mirascript/helpers/convert/to_number.py
from typing_extensions import TypeVar, overload
import math
import re

from ...vm.error import VmError
from ...vm.types.types import Uninitialized


def _parse_number(s: str) -> "float | None":
    if s == "":
        return None
    ch = s[0]
    if ch < "0" or ch > "9":
        return None

    if re.match(r"^0[bB][01]+$", s):
        return float(int(s[2:], 2))
    elif re.match(r"^0[oO][0-7]+$", s):
        return float(int(s[2:], 8))
    elif re.match(r"^0[xX][0-9a-fA-F]+$", s):
        return float(int(s[2:], 16))
    elif re.match(r"^[0-9][0-9]*$", s):
        return float(int(s, 10))
    elif re.match(r"^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$", s):
        return float(s)
    else:
        return None


def _string_to_number(s: str) -> "float | None":
    s = s.strip()
    if s == "":
        return None
    if s == "inf" or s == "+inf" or s == "Infinity" or s == "+Infinity":
        return math.inf
    if s == "-inf" or s == "-Infinity":
        return -math.inf
    if s == "nan" or s == "NaN":
        return math.nan
    if s.startswith("-"):
        num = _parse_number(s[1:])
        if num is not None:
            return -num
        return None
    elif s.startswith("+"):
        num = _parse_number(s[1:])
        if num is not None:
            return num
        return None
    else:
        return _parse_number(s)


T = TypeVar("T")


@overload
def toNumber(value) -> float: ...
@overload
def toNumber(value, fallback: T) -> "float | T": ...
def toNumber(value, fallback: T = Uninitialized) -> "float | T":
    if isinstance(value, bool):
        return 1.0 if value else 0.0

    if isinstance(value, (int, float)):
        return float(value)

    if isinstance(value, str):
        num = _string_to_number(value)
        if num is not None:
            return num

    if fallback is Uninitialized:
        raise VmError(f"Failed to convert value to number: {value}", math.nan)

    return fallback
